docs/DafnyRef/UserGuide.md

```diff
@@ -51,129 +96,315 @@ verification failure in a file that is being checked.
-`dafny run` takes only one .dfy file. Any other .dfy file and any other
-command-line argument that is not an option or its value or is after a `--`
-argument is considered a command-line argument for the user program being 
-built and run. For example,
+
+A guide to controlling and aiding the verification process is given in [a later section](#sec-verification)
+
+Various options control the verification process, in addition to all those described for [`dafny resolve`](#sec-dafny-resolve).
+
+- What is verified
+   - `--verify-included-files` (when enabled, all included files are verified, except library files, otherwise just those files on the command-line)
+   - `--relax-definite-assignment`
+   - `--track-print-effects`
+   - `--disable-nonlinear-arithmetic`
+
+- Control of the proof engine
+   - `--verification-time-limit`
+   - `--boogie`
+   - `--boogie-filter`
+
+
+#### 25.5.1.4. `dafny translate` {#sec-dafny-translate}
+
+The `dafny translate` command translates Dafny source code to source code for another target programming language.
+The command always performs the actions of `dafny resolve` and by default does the actions of `dafny verify`.
+
+In addition to generating the target source code, `dafny` may generate build artifacts to assist in compiling the generated code.
+The specific files generated depend on the target programming language.
+More detail is given in [the section on compilation](#sec-compilation).
+
+The `dafny` tool intends that the compiled program in the target language be a semantically faithful rendering of the 
+(verified) Dafny program. However, resource and language limitations make this not always possible. For example, Dafny
+can express and reason about arrays of unbounded size, not all target programming languages can represent arrays larger than the maximum signed 32-bit integer.
+
+Various options control the translation process, in addition to all those described for [`dafny resolve`](#sec-dafny-resolve) and [`dafny verify`](#sec-dafny-verify).
+
+- General options:
+   - `--no-verify` --- turns off all attempts to verify the program
+   - `--compile-verbose` --- print information about generated files
+
+- The translation target
+   - `--target` (or `-t`) --- the target language: cs (C#), java, js (Javascript), go, py (Python), cpp (C++)
+   - `--output` (or `-o`) --- location (file or folder depending on the target) of the generated file(s)
+   - `--include-runtime` --- include the Dafny runtime as source or a library in the target language
+   - `--optimize-erasable-datatype-wrapper`
+   - `--enforce-determinism`
+
+#### 25.5.1.5. `dafny build` {#sec-dafny-build}
+
+The `dafny build` command runs `dafny translate` and then compiles the result into an executable artifact for the target platform,
```

Comment:
I think I would prefer the default behavior of `dafny build` to be to produce a platform dependent executable, like a .EXE for Windows, so it would be a command that is by default agnostic to the execution platform (.NET, Java, etc.). `--target` should be a corner-case option, to use in case you want to implement `{:extern}` methods in a language other than C#.

`dafny translate` on the other hand should have `--target` not as an option but as an argument. I will change that today. Maybe we could add a bool option to `dafny translate`, `--compile-target` that will let Dafny attempt to invoke the target language compiler. Alternatively we could let the targets be subcommands of `dafny translate`, that optionally have the `--compile-target` option if that makes sense for that language (so yes for C#, no for JavaScript)

Reply:
As discussed elsewhere - I'm OK with the argument for translate and have adapted the text here accordingly.
I'd prefer to keep `dafny build` as a separate command from `dafny translate`. The compilation it does is useful now and the kind of artifact produced (library/executable, .dll, .jar, executable jar, jar with embedded JVM) can be options to build, even if they are not all yet implemented. Adding '--compile' to translate seems to me a step backwards from adoption of commands.